src/wpublish/QueriesAndChunks.cc

```diff
@@ -489,105 +499,168 @@ QueriesAndChunks::ScanTableSumsMap QueriesAndChunks::_calcScanTableSums() {
+                    LOGS(_log, LOG_LVL_DEBUG,
+                         __func__ << " b booted uq=" << countQId.qId << ":count=" << countQId.count
+                                  << " tasksBooted=" << tasksBooted
+                                  << " uqBootedTasksSum=" << uqBootedTasksSum);
+                }
+            }
         }
     }
-    return false;
+    return;
 }
 
-/// @return true if all Tasks for this query are complete.
-/// Precondition, _qStatsMtx must be locked.
-bool QueryStatistics::_isMostlyDead() const { return _tasksCompleted >= _size; }
+bool QueriesAndChunks::_bootUserQuery(QueryStatistics::Ptr queryToBoot,
+                                      wsched::BlendScheduler::Ptr const& bSched) {
+    // Get query info
+    if (queryToBoot == nullptr) {
+        LOGS(_log, LOG_LVL_ERROR, __func__ << " queryToBoot is nullptr");
+        // This really should never happen.
+        return false;
+    }
 
-ostream& operator<<(ostream& os, QueryStatistics const& q) {
-    lock_guard<mutex> gd(q._qStatsMtx);
-    os << QueryIdHelper::makeIdStr(q.queryId) << " time=" << q._totalTimeMinutes << " size=" << q._size
-       << " tasksCompleted=" << q._tasksCompleted << " tasksRunning=" << q._tasksRunning
-       << " tasksBooted=" << q._tasksBooted;
-    return os;
-}
+    // If the UserQuery was already booted, just add its booted `Task`s to the sum.
+    if (queryToBoot->_queryBooted) {
+        return true;
+    }
 
-void QueryStatistics::addTaskTransmit(double timeSeconds, int64_t bytesTransmitted, int64_t rowsTransmitted,
-                                      double bufferFillSecs) {
-    _histTimeTransmittingPerTask->addEntry(timeSeconds);
-    _histRowsPerTask->addEntry(rowsTransmitted);
-    _histSizePerTask->addEntry(bytesTransmitted);
-    _histTimeBufferFillPerTask->addEntry(bufferFillSecs);
-}
+    bool queryWasBooted = false;
+    // It may be possible for Tasks associated with a QueryId to be on more than 1 scheduler.
+    QueryStatistics::SchedTasksInfoMap schedTaskMap = queryToBoot->getSchedulerTasksInfoMap();
+    for (auto const& [key, schedInfo] : schedTaskMap) {
+        wsched::SchedulerBase::Ptr schedTarget = schedInfo.scheduler.lock();
+        if (schedTarget == nullptr) {
+            LOGS(_log, LOG_LVL_ERROR, __func__ << " schedTarg was nullptr for key=" << key);
+            continue;
+        }
+        if (bSched->isScanSnail(schedTarget)) {
+            LOGS(_log, LOG_LVL_DEBUG, __func__ << " schedTarg was snailScan for key=" << key);
+            continue;
+        }
 
-void QueryStatistics::addTaskRunQuery(double runTimeSeconds, double subchunkRunTimeSeconds) {
-    _histTimeRunningPerTask->addEntry(runTimeSeconds);
-    _histTimeSubchunkPerTask->addEntry(subchunkRunTimeSeconds);
+        LOGS(_log, LOG_LVL_INFO, "Booting uq queryId=" << queryToBoot->queryId);
+        queryToBoot->_queryBooted = true;
+        queryToBoot->_queryBootedTime = CLOCK::now();
```

Comment:
Is the current context modifying private data members of some other object? Is another object aware that its state could be modified externally w/o calling any public methods? This doesn't look safe.